specs/date_range_spec.rb

```diff
@@ -0,0 +1,191 @@
+require_relative 'spec_helper'
+
+describe 'DateRange' do
+  before do
+    @before = Date.new(2017, 10, 1)
+    @after = Date.new(2017, 10, 14)
+  end
+
+  describe 'self.range_to' do
+    it 'returns an array of Dates' do
+      dates = DateRange.range_to(@before, @after)
+
+      dates.must_be_kind_of Array
+
+      dates.each do |date|
+        date.must_be_kind_of Date
+      end
+    end
+
+    it 'includes all dates up to, but not including, the end date' do
+      dates = DateRange.range_to(@before, @after)
+
+      dates.first.strftime.must_equal '2017-10-01'
+      dates.last.strftime.must_equal '2017-10-13'
+    end
+
+    it 'raises DatesError if start date is after end date' do
+      proc {
+        DateRange.range_to(@after, @before)
+      }.must_raise DatesError
+    end
+
+    it 'raises DatesError if dates do not span at least 1 night' do
+      proc {
+        DateRange.range_to(@before, @before)
+      }.must_raise DatesError
+    end
+
+    it 'raises ArgumentError if passed invalid dates' do
+      proc {
+        DateRange.range_to('sea', 'HAWKS')
+      }.must_raise ArgumentError
+    end
+  end
+
+  describe 'self.overlap?' do
+    it 'returns true if provided date range overlaps' do
+      overlap = DateRange.overlap?('2017-09-06', '2017-09-07', '2017-09-06', '2017-09-07')
+      overlap.must_equal true
+    end
+
+    it 'returns false if provided date range does not overlap' do
+      overlap = DateRange.overlap?('2017-10-14', '2017-10-15', '2017-09-06', '2017-09-07')
+      overlap.must_equal false
+    end
+
+    it 'returns true for partial overlap' do
+      overlap = DateRange.overlap?('2017-09-06', '2017-09-10', '2017-09-8', '2017-09-20')
+      overlap.must_equal true
+    end
+
+    it 'allows one range to end the day the next starts' do
+      overlap = DateRange.overlap?('2017-09-06', '2017-09-10', '2017-09-10', '2017-09-20')
+      overlap.must_equal false
+    end
```

Comment:
I like these test cases! There are many similar cases it might be interesting to explore here around whether or not a room is available. My entire list includes:
- Does Overlap:
  - Same dates
  - Overlaps in the front
  - Overlaps in the back
  - Completely contained
  - Completely containing
- Does Not Overlap:
  - Completely before
  - Completely after
  - Ends on the checkin date
  - Starts on the checkout date